File: src/main/java/me/zeroeightsix/kami/command/SettingValueArgumentType.java
package me.zeroeightsix.kami.command;

import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.SuggestionsBuilder;
import me.zeroeightsix.kami.setting.Setting;
import net.minecraft.text.LiteralText;

import java.util.concurrent.CompletableFuture;

public class SettingValueArgumentType extends DependantArgumentType<String, Setting> {

    public static final DynamicCommandExceptionType INVALID_VALUE_EXCEPTION = new DynamicCommandExceptionType((object) -> new LiteralText("Invalid value '" + ((Object[]) object)[0] + "' for property '" + ((Object[]) object)[1] + "'"));

    public SettingValueArgumentType(ArgumentType<Setting> dependantType, String dependantArgument, int shift) {
        super(dependantType, dependantArgument, shift);
    }

    public static SettingValueArgumentType value(ArgumentType<Setting> dependantType, String dependantArgument, int shift) {
        return new SettingValueArgumentType(dependantType, dependantArgument, shift);
    }

    @Override
    public String parse(StringReader reader) throws CommandSyntaxException {
        Setting setting = findDependencyValue(reader);
        String string = reader.readUnquotedString();
        try {
            Object v = setting.convertFromString(string);
            return string;
        } catch (Exception ignored) {}
        throw INVALID_VALUE_EXCEPTION.create(new Object[] { string, setting.getName() });
    }

    @Override
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> context, SuggestionsBuilder builder) {
        return findDependencyValue(context, Setting.class).listSuggestions(context, builder);
    }
}

File: src/main/kotlin/me/zeroeightsix/kami/feature/command/SettingValueArgumentType.kt
package me.zeroeightsix.kami.feature.command

import com.mojang.brigadier.StringReader
import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import com.mojang.brigadier.suggestion.Suggestions
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import io.github.fablabsmc.fablabs.api.fiber.v1.FiberId
import io.github.fablabsmc.fablabs.api.fiber.v1.tree.ConfigLeaf
import me.zeroeightsix.kami.setting.SettingAnnotationProcessor
import me.zeroeightsix.kami.setting.SettingInterface
import net.minecraft.text.LiteralText
import java.util.concurrent.CompletableFuture
import java.util.function.Function

class SettingValueArgumentType(
    dependantType: ArgumentType<ConfigLeaf<*>>,
    dependantArgument: String,
    shift: Int
) : DependantArgumentType<String, ConfigLeaf<*>>(
    dependantType,
    dependantArgument,
    shift
) {
    @Throws(CommandSyntaxException::class)
    override fun parse(reader: StringReader): String {
        val setting = findDependencyValue(reader)
        val string = reader.readUnquotedString()
        return if (setting.getInterface().canFromString(string)) {
            string
        } else {
            throw INVALID_VALUE_EXCEPTION.create(
                arrayOf<Any>(
                    string,
                    setting.name
                )
            )
        }
    }

    override fun <S> listSuggestions(
        context: CommandContext<S>,
        builder: SuggestionsBuilder
    ): CompletableFuture<Suggestions> {
        return findDependencyValue(context, ConfigLeaf::class.java).getInterface().listSuggestions(
            context,
            builder
        )
    }

    companion object {
        val INVALID_VALUE_EXCEPTION =
            DynamicCommandExceptionType(Function { `object`: Any ->
                LiteralText(
                    "Invalid value '" + (`object` as Array<Any>)[0] + "' for property '" + `object`[1] + "'"
                )
            })

        fun value(
            dependantType: ArgumentType<ConfigLeaf<*>>,
            dependantArgument: String,
            shift: Int
        ): SettingValueArgumentType {
            return SettingValueArgumentType(dependantType, dependantArgument, shift)
        }
    }
}

fun <T> ConfigLeaf<T>.getInterface(): SettingInterface<T> =
    this.getAttributeValue(FiberId("kami", "setting_interface"), SettingAnnotationProcessor.INTERFACE_TYPE).get() as SettingInterface<T>